samples/nrf_desktop/src/modules/hid_forward.c

```diff
@@ -0,0 +1,220 @@
+
+	u16_t x = sys_get_le16(x_buff);
+	u16_t y = sys_get_le16(y_buff);
+
+	if (x > REPORT_MOUSE_XY_MAX) {
+		x |= 0xF000;
+	}
+	if (y > REPORT_MOUSE_XY_MAX) {
+		y |= 0xF000;
+	}
+
+	struct hid_mouse_event *event;
+
+	u32_t key = irq_lock();
+
+	if (next_event) {
+		__ASSERT_NO_MSG(usb_busy);
+
+		LOG_WRN("Event override");
+		event = next_event;
+	} else {
+		event = new_hid_mouse_event();
+	}
+
+	event->subscriber = usb_id;
+
+	event->button_bm = button_bm;
+	event->wheel     = wheel;
+
+	event->dx = x;
+	event->dy = y;
+
+	if (!usb_busy) {
+		EVENT_SUBMIT(event);
+		usb_busy = true;
+	} else {
+		next_event = event;
+	}
+
+	irq_unlock(key);
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static void hidc_ready(struct bt_gatt_hids_c *hids_c)
+{
+	int err = bt_gatt_hids_c_rep_subscribe(&hidc, hidc.rep_info[0],
+					       hidc_read);
+
+	if (err) {
+		LOG_ERR("Cannot subscribe to report (err:%d)", err);
+	}
+}
+
+static void hidc_pm_update(struct bt_gatt_hids_c *hids_c)
+{
+	LOG_INF("Protocol mode updated");
+}
+
+static void hidc_prep_error(struct bt_gatt_hids_c *hids_c, int err)
+{
+	if (err) {
+		LOG_ERR("err:%d", err);
+	}
+}
+
+static void init(void)
+{
+	static const struct bt_gatt_hids_c_init_params params = {
+		.ready_cb = hidc_ready,
+		.prep_error_cb = hidc_prep_error,
+		.pm_update_cb = hidc_pm_update,
+	};
+
+	bt_gatt_hids_c_init(&hidc, &params);
+}
+
+static int assign_handles(struct bt_gatt_dm *dm)
+{
+	int err = bt_gatt_hids_c_handles_assign(dm, &hidc);
+
+	if (err) {
+		LOG_ERR("Cannot assign handles (err:%d)", err);
+	}
+
+	return err;
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (is_hid_report_sent_event(eh)) {
+		u32_t key = irq_lock();
+		if (next_event) {
+			EVENT_SUBMIT(next_event);
+			next_event = NULL;
+		} else {
+			usb_busy = false;
+		}
+		irq_unlock(key);
+
+		return false;
+	}
+
+	if (is_module_state_event(eh)) {
+		const struct module_state_event *event =
+			cast_module_state_event(eh);
+
+		if (check_state(event, MODULE_ID(ble_state),
+				MODULE_STATE_READY)) {
+			static bool initialized;
+
+			__ASSERT_NO_MSG(!initialized);
+			initialized = true;
+
+			init();
+			module_set_state(MODULE_STATE_READY);
+		}
+
+		return false;
+	}
+
+	if (is_ble_discovery_complete_event(eh)) {
+		const struct ble_discovery_complete_event *event =
+			cast_ble_discovery_complete_event(eh);
+
+		assign_handles(event->dm);
+
+		return false;
+	}
+
+	if (is_hid_report_subscription_event(eh)) {
+		const struct hid_report_subscription_event *event =
+			cast_hid_report_subscription_event(eh);
+
+		if (event->subscriber == usb_id) {
+			atomic_set(&usb_ready, event->enabled);
+		}
+
+		return false;
+	}
+
+	if (is_usb_state_event(eh)) {
+		const struct usb_state_event *event =
+			cast_usb_state_event(eh);
+
+		switch (event->state) {
+		case USB_STATE_POWERED:
+			usb_id = event->id;
+			break;
+		case USB_STATE_DISCONNECTED:
+			usb_id = NULL;
```

Comment:
Shouldn't you swap the order of these 2 lines? In one corner case, you may generate hid_mouse_event with NULL usb_id if hidc_read event handler is called right after this assignment

Reply:
well spotted
Actually I kind of depended on the fact that if subscriber id is wrong it will be rejected on the other side ;)

Clean code would take a lock and keep local copies of usb_ready + usb_id. The thing is even then you can have a race as disconnection can happen right after you created a local copy (or after processing data). So anyway we need other side to reject the event correctly and (unless there is a bug) it should do it.